diff --git a/configs/s5p6818_nanopim3_defconfig b/configs/s5p6818_nanopim3_defconfig
index 0e221eb..f60ebc8 100644
--- a/configs/s5p6818_nanopim3_defconfig
+++ b/configs/s5p6818_nanopim3_defconfig
@@ -17,3 +17,4 @@ CONFIG_CMD_EXT4_IMG_WRITE=y
 CONFIG_CMD_SD_RECOVERY=y
 CONFIG_OF_CONTROL=y
 CONFIG_OF_EMBED=y
+CONFIG_SPL=y
diff --git a/include/configs/s5p6818_nanopim3.h b/include/configs/s5p6818_nanopim3.h
index d899d36..41335cf 100644
--- a/include/configs/s5p6818_nanopim3.h
+++ b/include/configs/s5p6818_nanopim3.h
@@ -100,7 +100,6 @@
  *	U-Boot Environments
  */
 /* refer to common/env_common.c	*/
-#define CONFIG_BOOTDELAY			3
 
 /*-----------------------------------------------------------------------
  * Miscellaneous configurable options
@@ -279,17 +278,82 @@
 /*-----------------------------------------------------------------------
  * ENV
  */
-#define CONFIG_BOOTARGS \
-    "console=ttySAC0,115200n8 initrd=0x49000000,0x400000 root=/dev/mmcblk1p1 console=tty1"
-
-#define CONFIG_EXTRA_ENV_SETTINGS	\
-			"fdt_high=0xffffffffffffffff"
-
-#define CONFIG_BOOTCOMMAND	\
-	"ext4load mmc 0:1 0x48000000 boot/Image; " \
-	"mw 0x49000000 0 0x400000; " \
-	"ext4load mmc 0:1 0x49000000 boot/initrd.img; " \
-	"ext4load mmc 0:1 0x4a000000 boot/s5p6818-nanopi-m3.dtb; " \
-	"booti 0x48000000 - 0x4a000000"
-
-#endif /* __CONFIG_H__ */
+#define ENV_DEVICE_SETTINGS \
+		"stdin=serial,lcd\0" \
+	"stdout=serial,lcd\0" \
+	"stderr=serial,lcd\0"
+
+#define ENV_MEM_LAYOUT_SETTINGS \
+    "ramdisk_addr_r=0x49000000\0" \
+    "kernel_addr_r=0x48000000\0" \
+    "scriptaddr=0x47000000\0" \
+    "fdtfile=nexell/s5p6818-nanopi-m3.dtb\0" \
+    "fdt_addr_r=0x4a000000\0" \
+    "fdt_high=0xffffffffffffffff"
+
+#define BOOTCMDS_MMC \
+		"mmc_boot=" \
+		"setenv devtype mmc; " \
+		"if mmc dev ${devnum}; then " \
+			"run scan_boot; " \
+		"fi\0" \
+	"bootcmd_mmc0=setenv devnum 0; run mmc_boot;\0"
+#define BOOT_TARGETS_MMC "mmc0"
+
+#define BOOTCMDS_COMMON \
+		"rootpart=1\0" \
+	\
+	"do_script_boot="                                                 \
+		"load ${devtype} ${devnum}:${rootpart} "                  \
+			"${scriptaddr} ${prefix}${script}; "              \
+		"source ${scriptaddr}\0"                                  \
+	\
+	"script_boot="                                                    \
+		"for script in ${boot_scripts}; do "                      \
+			"if test -e ${devtype} ${devnum}:${rootpart} "    \
+					"${prefix}${script}; then "       \
+				"echo Found ${prefix}${script}; "         \
+				"run do_script_boot; "                    \
+				"echo SCRIPT FAILED: continuing...; "     \
+			"fi; "                                            \
+		"done\0"                                                  \
+	\
+	"do_sysboot_boot="                                                \
+		"sysboot ${devtype} ${devnum}:${rootpart} any "           \
+			"${scriptaddr} ${prefix}extlinux/extlinux.conf\0" \
+	\
+	"sysboot_boot="                                                   \
+		"if test -e ${devtype} ${devnum}:${rootpart} "            \
+				"${prefix}extlinux/extlinux.conf; then "  \
+			"echo Found ${prefix}extlinux/extlinux.conf; "    \
+			"run do_sysboot_boot; "                           \
+			"echo SCRIPT FAILED: continuing...; "             \
+		"fi\0"                                                    \
+	\
+	"scan_boot="                                                      \
+		"echo Scanning ${devtype} ${devnum}...; "                 \
+		"for prefix in ${boot_prefixes}; do "                     \
+			"run sysboot_boot; "                              \
+			"run script_boot; "                               \
+		"done\0"                                                  \
+	\
+	"boot_targets=" \
+		BOOT_TARGETS_MMC " " \
+		"\0" \
+	\
+	"boot_prefixes=/\0" \
+	\
+	"boot_scripts=boot.scr.uimg\0" \
+	\
+	BOOTCMDS_MMC
+
+#define CONFIG_BOOTCOMMAND \
+		"for target in ${boot_targets}; do run bootcmd_${target}; done"
+
+#define CONFIG_EXTRA_ENV_SETTINGS \
+	ENV_DEVICE_SETTINGS \
+	ENV_MEM_LAYOUT_SETTINGS \
+	BOOTCMDS_COMMON
+
+#include <config_distro_defaults.h>
+#endif
